libexec/rtld-elf/rtld_c18n.h

```diff
@@ -0,0 +1,159 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#ifndef RTLD_C18N_H
+#define RTLD_C18N_H
+
+/*
+ * Global symbols
+ */
+extern uintptr_t sealer_pltgot, sealer_jmpbuf, sealer_tramp;
+extern const char *ld_compartment_utrace;
+extern const char *ld_compartment_enable;
+extern const char *ld_compartment_overhead;
+
+void tramp_init(void);
+
+/*
+ * libthr support
+ */
+extern void (*thr_thread_start)(struct pthread *);
+extern void (*thr_sighandler)(int, siginfo_t *, void *);
+
+/*
+ * Policies
+ */
+#define	C18N_RTLD_COMPARTMENT_ID	1
+
+typedef uint16_t compart_id_t;
+
+struct compart {
+	/*
+	 * Name of the compartment
+	 */
+	const char *name;
+	/*
+	 * NULL-terminated array of libraries that belong to the compartment
+	 */
+	const char **libraries;
+};
+
+struct policy {
+	struct compart *coms;
+	size_t count;
+};
+
+void tramp_add_comparts(struct policy *);
+compart_id_t compart_id_allocate(const char *);
+
+/*
+ * Stack switching
+ */
+#define	C18N_STACK_TABLE_SIZE	2
+
+struct tramp_stk_table_info {
+	size_t capacity;
+	size_t stack_sizes[];
+};
+
+struct tramp_stk_table {
+	void (*resolver)(void);
+	struct tramp_stk_table_info *info;
+	void *stacks[];
+};
+
+struct Struct_Stack_Entry {
+    SLIST_ENTRY(Struct_Stack_Entry) link;
+    void *stack;
+};
+
+void _rtld_get_rstk(void);
+struct tramp_stk_table *tramp_stk_table_expand(struct tramp_stk_table *, size_t);
+
+/*
+ * Trampolines
+ */
+#define	DT_CHERI_C18N_SIG	0x60000100	/* function signature (c18n) */
+
+typedef uint8_t tramp_sig_int;
+
+/* Must not be reordered */
+enum tramp_ret_args: unsigned char {
+	TWO,
+	ONE,
+	NONE,
+	INDIRECT
+};
+
+struct tramp_sig {
+	enum tramp_ret_args ret_args : 2;
+	bool mem_args : 1;
+	unsigned char reg_args : 4;
+	bool valid : 1;
+};
+
+struct tramp_data {
+	void *target;
+	const Obj_Entry *defobj;
+	const Elf_Sym *def;
+	void *entry;
+	struct tramp_sig sig;
+};
+_Static_assert(sizeof(struct tramp_sig) == sizeof(tramp_sig_int),
+    "Unexpected tramp_sig size");
+
+void tramp_hook(int, void *, const Obj_Entry *, const Elf_Sym *, void *, void *);
+size_t tramp_compile(char **, const struct tramp_data *);
+void *tramp_intern(const Obj_Entry *reqobj, const struct tramp_data *);
+struct tramp_sig tramp_fetch_sig(const Obj_Entry *, unsigned long);
+
+static inline long
+tramp_sig_to_otype(struct tramp_sig sig)
+{
+	return (sig.ret_args | (sig.mem_args << 2) | (sig.reg_args << 3));
+}
+
+static inline tramp_sig_int
+tramp_sig_to_int(struct tramp_sig sig)
+{
+	tramp_sig_int ret;
+	memcpy(&ret, &sig, sizeof(tramp_sig_int));
+	return ret;
```

Comment:
```suggestion

	memcpy(&ret, &sig, sizeof(tramp_sig_int));
	return (ret);
```

Reply:
Though I question if this function is a good idea (and its type); it's only used for error messages, which seem like they would be better off just printing the fields